### PR TITLE
Add support for `arith_offset`

### LINKED ISSUE #227: Add support for `arith_offset`
We can't translate the following program:

```rust
fn main() {
    let ptr = &mut 42u8 as *mut u8;
    let _ = ptr.wrapping_add(1);
}
```

**Error message**:
```
thread 'rustc' panicked at minimize/src/bb.rs:288:25:
unsupported Rust intrinsic `arith_offset`
```

### PR DESCRIPTION
**Pull request by sun-jacobi**
Closes #227

### CODE REVIEW

File: spec/mem/basic.md [hunk 1]
```diff
@@ -118,7 +118,7 @@ impl<Provenance, Extra> Allocation<Provenance, Extra> {
         }
 
         let offset_in_alloc = addr - self.addr;
-        if offset_in_alloc < 0 || offset_in_alloc + pointee_size.bytes() > self.size().bytes() {
+        if !pointee_size.is_zero() && (offset_in_alloc < 0 || offset_in_alloc + pointee_size.bytes() > self.size().bytes()) {
```
Comment (RalfJung) [inline spec/mem/basic.md:121]: Why this change? This function is never even called for zero-sized pointees, I think?

File: tooling/minimize/src/bb.rs [hunk 2]
```diff
@@ -10,6 +10,11 @@ enum StatementResult {
     Intrinsic { intrinsic: IntrinsicOp, destination: PlaceExpr, arguments: List<ValueExpr> },
 }
 
+struct TerminatorResult {
```
Comment (RalfJung) [inline tooling/minimize/src/bb.rs:13]:
```suggestion
/// A MIR terminator becomes a MiniRust terminator, possibly preceded by a list
/// of statements to execute at the end of the previous basic block.
struct TerminatorResult {
```

File: tooling/minimize/src/bb.rs [hunk 3]
```diff
@@ -229,27 +239,46 @@ impl<'cx, 'tcx> FnCtxt<'cx, 'tcx> {
...
+                        terminator: Terminator::Intrinsic {
+                            intrinsic: IntrinsicOp::RawEq,
+                            arguments: args
+                                .iter()
+                                .map(|x| self.translate_operand(&x.node, x.span))
+                                .collect(),
+                            ret: self.translate_place(&destination, span),
+                            next_block: target.as_ref().map(|t| self.bb_name_map[t]),
+                        },
+                        stmts: List::new(),
```
Comment (RalfJung) [inline tooling/minimize/src/bb.rs:265]: Please put `stmts` before `terminator`, since logically they also execute first.

File: tooling/minimize/src/bb.rs [hunk 4]
```diff
@@ -229,27 +239,46 @@ impl<'cx, 'tcx> FnCtxt<'cx, 'tcx> {
...
                     },
+                "arith_offset" => {
+                    let terminator =
+                        Terminator::Goto(target.as_ref().map(|t| self.bb_name_map[t]).unwrap());
```
Comment (RalfJung) [inline tooling/minimize/src/bb.rs:269]:
```suggestion
                        Terminator::Goto(self.bb_name_map[target.unwrap()]);
```

File: tooling/minimize/src/bb.rs [hunk 5]
```diff
@@ -229,27 +239,46 @@ impl<'cx, 'tcx> FnCtxt<'cx, 'tcx> {
...
+                    let destination = self.translate_place(&destination, span);
+
+                    let source = ValueExpr::BinOp {
```
Comment (RalfJung) [inline tooling/minimize/src/bb.rs:272]: "source" is a confusing name here, maybe call it "result" or so?

File: tooling/minimize/tests/pass/arith_offset.rs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+//@[tree]compile-flags: --minimize-tree-borrows
+
+fn wrapping_offset() {
+    let data = [1u8, 2, 3, 4, 5];
+    let first = &data[0] as *const u8;
+
+    unsafe {
+        assert!(*first == 1);
+        assert!(*first.wrapping_offset(2) == 3);
+        assert!(*first.wrapping_offset(4) == 5);
```
Comment (RalfJung) [inline tooling/minimize/tests/pass/arith_offset.rs:11]: Please also add a case that actually goes out-of-bounds. Something like `*first.wrapping_offset(100).wrapping_offset(-100) == 1`.

File: spec/mem/basic.md (hunk 1 above)
Comment (RalfJung) [reply spec/mem/basic.md:121]: Ah, I see, this is also used directly by TB. I think that should be done differently... I have pushed 6e77e25 which avoids calling `offset_in_alloc` when the size is zero, then no change should be needed here.

Comment (RalfJung) [reply spec/mem/basic.md:121]: Sorry, the push failed... it's now pushed as commit 5e38d1c.

**Comment by sun-jacobi** [comment]
Resolved above.

**Comment by RalfJung** [comment]
Thanks :)
